File: Antelope/Antelope.cpp
#include "Antelope.h"
#include "../Settings.h"

Antelope::Antelope(World *world, int x, int y, int parent1Id, int parent2Id) :
    Animal(ANTELOPE_STRENGTH, ANTELOPE_INITIATIVE, ANTELOPE_SYMBOL, ANTELOPE_COLOR, world, x, y, parent1Id, parent2Id) {}

void Antelope::action() {
  Animal::action();
  Animal::action();
}

void Antelope::collision(Organism *organism) {
  Animal::collision(organism);
}

Organism* Antelope::createChild(int x, int y, int parent1Id, int parent2Id) {
  return new Antelope(this->getWorld(), x, y, parent1Id, parent2Id);
}

std::string Antelope::getType() const { return "Antelope"; }

bool Antelope::deflectedAttack(Organism *attacker) {
  int deflectChance = rand() % 2;
  if (deflectChance == 0) {
    Position newPosition = this->getWorld()->generateNewSafePosition(attacker, this->getPosition().getX(), this->getPosition().getY());
    this->setPosition(newPosition.getX(), newPosition.getY());
    this->getWorld()->addLog("%s POS:(%d, %d) run away from %s POS:(%d, %d)", this->getType().c_str(), this->getPreviousX(), this->getPreviousY(), attacker->getType().c_str(), attacker->getPreviousX(), attacker->getPreviousY());
    return true;
  }

  return Animal::deflectedAttack(attacker);
}

File: src/Antelope/Antelope.java
package Antelope;

import Animal.Animal;
import Field.Field;
import Organism.Organism;
import Organism.OrganismData;
import Settings.Settings;
import World.World;

import java.util.Random;

public class Antelope extends Animal {

    public Antelope(World world, Field field, int parent1Id, int parent2Id) {
        super(new OrganismData(world.getNextOrganismId(), Settings.ANTELOPE_INITIATIVE, Settings.ANTELOPE_STRENGTH, Settings.ANTELOPE_IMAGE, parent1Id, parent2Id), world, field);
    }

    @Override
    public void action() {
        super.action();
        super.action();
    }

    @Override
    public String getType() {
        return "Antelope";
    }

    @Override
    public void reproduce(Field field, int parent1Id, int parent2Id) {
        Organism antelope = new Antelope(this.world, field, parent1Id, parent2Id);
        this.world.addOrganism(antelope);
    }

    @Override
    public boolean deflectedAttack(Organism attacker) {
        Random rand = new Random();
        int deflectionChance = rand.nextInt(2);
        if (deflectionChance == 0) {
            Field field = this.world.getFreeNeighbourField(this.currentField);
            if (field != null) {
                this.currentField = field;
            }
            return true;
        }

        return super.deflectedAttack(attacker);
    }
}
